lib/ecmascript.ts

```diff
@@ -2095,1136 +1976,1397 @@ export const ES = ObjectAssign({}, ES2020, {
+  );
+  return result;
+}
+
+export function GetFormatterParts(timeZone, epochMilliseconds) {
+  const formatter = getIntlDateTimeFormatEnUsForTimeZone(timeZone);
+  // FIXME: can this use formatToParts instead?
+  const datetime = formatter.format(new Date(epochMilliseconds));
+  const [date, fullYear, time] = datetime.split(/,\s+/);
+  const [month, day] = date.split(' ');
+  const [year, era] = fullYear.split(' ');
+  const [hour, minute, second] = time.split(':');
+  return {
+    year: era === 'BC' ? -year + 1 : +year,
+    month: +month,
+    day: +day,
+    hour: hour === '24' ? 0 : +hour, // bugs.chromium.org/p/chromium/issues/detail?id=1045791
+    minute: +minute,
+    second: +second
+  };
+}
+
+export function GetIANATimeZoneEpochValue(
+  id,
+  year,
+  month,
+  day,
+  hour,
+  minute,
+  second,
+  millisecond,
+  microsecond,
+  nanosecond
+) {
+  const ns = GetEpochFromISOParts(year, month, day, hour, minute, second, millisecond, microsecond, nanosecond);
+  if (ns === null) throw new RangeError('DateTime outside of supported range');
+  const dayNanos = bigInt(DAYMILLIS).multiply(1e6);
+  let nsEarlier = ns.minus(dayNanos);
+  if (nsEarlier.lesser(NS_MIN)) nsEarlier = ns;
+  let nsLater = ns.plus(dayNanos);
+  if (nsLater.greater(NS_MAX)) nsLater = ns;
+  const earliest = GetIANATimeZoneOffsetNanoseconds(nsEarlier, id);
+  const latest = GetIANATimeZoneOffsetNanoseconds(nsLater, id);
+  const found = earliest === latest ? [earliest] : [earliest, latest];
+  return found
+    .map((offsetNanoseconds) => {
+      const epochNanoseconds = bigInt(ns).minus(offsetNanoseconds);
+      const parts = GetIANATimeZoneDateTimeParts(epochNanoseconds, id);
+      if (
+        year !== parts.year ||
+        month !== parts.month ||
+        day !== parts.day ||
+        hour !== parts.hour ||
+        minute !== parts.minute ||
+        second !== parts.second ||
+        millisecond !== parts.millisecond ||
+        microsecond !== parts.microsecond ||
+        nanosecond !== parts.nanosecond
+      ) {
+        return undefined;
       }
-    } while (isOverflow);
-    return { days, nanoseconds, dayLengthNs: MathAbs(dayLengthNs) };
-  },
-  BalanceDuration: (
-    days,
-    hours,
-    minutes,
-    seconds,
-    milliseconds,
-    microseconds,
-    nanoseconds,
-    largestUnit,
-    relativeTo = undefined
-  ) => {
-    if (ES.IsTemporalZonedDateTime(relativeTo)) {
-      const endNs = ES.AddZonedDateTime(
-        GetSlot(relativeTo, INSTANT),
-        GetSlot(relativeTo, TIME_ZONE),
-        GetSlot(relativeTo, CALENDAR),
-        0,
-        0,
-        0,
-        days,
-        hours,
-        minutes,
-        seconds,
-        milliseconds,
-        microseconds,
-        nanoseconds
-      );
-      const startNs = GetSlot(relativeTo, EPOCHNANOSECONDS);
-      nanoseconds = endNs.subtract(startNs);
+      return epochNanoseconds;
+    })
+    .filter((x) => x !== undefined);
+}
+
+export function LeapYear(year) {
```

Comment:
Is there an easy way to determine whether all of the exported functions in the old ES object are called in any other files?  This function is called (twice in calendar.ts) but I suspect that at least one ecmasript.ts function probably snuck into the ES object while only being used inside ecmascript.ts itself.

Reply:
SGTM. Try this: `npx ts-unused-exports ./tsconfig.json lib/ecmascript.ts`. 

If that doesn't work, there's always manual brute force + muscle memory: 

1. In VSCode, open all the other /lib/*.ts files in the editor, so that TS will find errors in them without having to run anything.
2. Search for `export ` in ecmascript.mjs
3. Hit the Delete key.
4. Are there errors showing up in VSCode's "problems" pane?  If not, use Find All References to verify that the function is unused, and if so then delete it!
5. If not, then Cmd+Z to undo
6. Cmd+G to advance to the next match
7. GOTO 3

This is manual but after the first few iterations it will sink into muscle memory movements so probably could get it done in 10-20 (boring) minutes. 😄

Reply:
Turns out that `npx ts-prune` works perfectly. See #62 and #63.

Reply:
And we can even make ts-prune into a CI test to detect future unused functions or unused exports. I put this CI test into #62.